internal/modulespecifiers/specifiers.go

```diff
@@ -644,6 +659,54 @@ func tryGetModuleNameAsNodeModule(
 	overrideMode core.ResolutionMode,
 ) string {
 	parts := getNodeModulePathParts(pathObj.FileName)
+
+	// TODO understand what feature this part impacts
+	pnpPackageName := ""
+
+	pnpApi := pnp.GetPnpApi(importingSourceFile.FileName())
+	if pnpApi != nil {
+		fromLocator, _ := pnpApi.FindLocator(importingSourceFile.FileName())
+		toLocator, _ := pnpApi.FindLocator(pathObj.FileName)
+
+		// Don't use the package name when the imported file is inside
+		// the source directory (prefer a relative path instead)
+		if fromLocator == toLocator {
+			return ""
+		}
+
+		if fromLocator != nil && toLocator != nil {
+			fromInfo := pnpApi.GetPackage(fromLocator)
+
+			useToLocator := false
+
+			for i := range fromInfo.PackageDependencies {
+				isAlias := fromInfo.PackageDependencies[i].IsAlias()
+				if isAlias && fromInfo.PackageDependencies[i].AliasName == toLocator.Name && fromInfo.PackageDependencies[i].Reference == toLocator.Reference {
+					useToLocator = true
+					break
+				} else if fromInfo.PackageDependencies[i].Ident == toLocator.Name && fromInfo.PackageDependencies[i].Reference == toLocator.Reference {
+					useToLocator = true
+					break
+				}
+			}
+
+			if useToLocator {
+				pnpPackageName = toLocator.Name
+			}
+		}
+
+		if parts != nil {
+			toInfo := pnpApi.GetPackage(toLocator)
+			parts = &NodeModulePathParts{
+				TopLevelNodeModulesIndex: -1,
+				TopLevelPackageNameIndex: -1,
+				// The last character from packageLocation is the trailing "/", we want to point to it
+				PackageRootIndex: len(toInfo.PackageLocation) - 1,
+				FileNameIndex:    strings.LastIndex(pathObj.FileName, "/"),
+			}
+		}
+	}
```

Comment:
I think this should be moved into a separate function, and be called with an early return. Something like this:

```
module_name = tryGetModuleNameAsPnpPackage(...)
if module_name != nil {
  return module_name
}
```